File: 2022/day05.py
import re
from itertools import zip_longest

with open("input_files/day05", "r") as f:
    crates, moves = f.read().split("\n\n")

    crates = crates.splitlines()[:-1]
    crates = [crate[1::4] for crate in crates]
    crates = [*zip_longest(*crates[::-1], fillvalue=" ")]
    crates = [[e for e in crate if not e.isspace()] for crate in crates]

    moves = re.findall(r"move (\d+) from (\d+) to (\d+)", moves)
    moves = [map(int, move) for move in moves]
    moves = [[amt, from_ - 1, to - 1] for amt, from_, to in moves]

crate_mover_9000 = [crate[:] for crate in crates]
crate_mover_9001 = [crate[:] for crate in crates]

for amt, from_, to in moves:
    crate_mover_9000[to] += crate_mover_9000[from_][:-amt - 1:-1]
    del crate_mover_9000[from_][-amt:]

    crate_mover_9001[to] += crate_mover_9001[from_][-amt:]
    del crate_mover_9001[from_][-amt:]

print("".join(crate[-1] for crate in crate_mover_9000))
print("".join(crate[-1] for crate in crate_mover_9001))

File: kotlin/src/main/kotlin/y2022/day05.kt
package y2022

import read
import transpose

fun main() {
    val (crateData, moveData) = read("y2022/day05").split("\n\n")

    val width = crateData.lines().maxOf { it.length }
    val crates = crateData.lines()
        .dropLast(1)
        .map {
            it.padEnd(width, ' ')
                .slice(1 until width step 4)
                .toList()
        }
        .transpose()
        .map { it.filter { char -> !char.isWhitespace() } }

    val moveRegex = "move (\\d+) from (\\d+) to (\\d+)".toRegex()
    val moves = moveRegex.findAll(moveData)
        .map { it.groupValues.drop(1)}
        .map { it.map { e -> e.toInt() }  }
        .map { (amt, from, to) -> Triple(amt, from - 1, to - 1) }

    val crateMover9000 = crates.map { it.toMutableList() }.toMutableList()
    val crateMover9001 = crates.map { it.toMutableList() }.toMutableList()

    for ((amt, from, to) in moves) {
        val toMoveSingle = crateMover9000[from].takeLast(amt).reversed()
        crateMover9000[to].addAll(toMoveSingle)
        crateMover9000[from] = crateMover9000[from].dropLast(amt).toMutableList()

        val toMoveMany = crateMover9001[from].takeLast(amt)
        crateMover9001[to].addAll(toMoveMany)
        crateMover9001[from] = crateMover9001[from].dropLast(amt).toMutableList()
    }

    println(crateMover9000.map { it.last() }.joinToString(""))
    println(crateMover9001.map { it.last() }.joinToString(""))
}
